driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1183,270 @@ int getNumPinnedToTransaction() {
+     * An {@code open} task blocked because of the max connecting limit is also being counted towards the max size limit,
+     * i.e., it may block a {@code create} task. If we were to execute both kinds of tasks in the same thread, then it would
+     * totally order all tasks, making it possible for a {@code create} task to block an {@code open} task and, thus, open opportunities for
+     * tasks deadlocking. As a result, we must maintain at least two sets of threads for two kinds of tasks.
+     * <p>
+     * Due to tasks having a timeout, we also must monitor queued tasks to determine which ones timed out and completing them accordingly.
+     * While this logic can be incorporated in {@code create}/{@code open} tasks, doing so would have significantly complicate them further,
+     * and it is especially not obvious how to maintain fairness in that case.
+     */
     @ThreadSafe
-    private static class Workers implements AutoCloseable {
-        private volatile ExecutorService getter;
-        private volatile ExecutorService opener;
+    private static class AsyncWorkManager implements AutoCloseable {
+        private volatile State state;
+        private volatile BlockingQueue<Task> gets;
+        private volatile BlockingQueue<Task> opens;
         private final Lock lock;
+        private final Condition closedOrTaskAddedCondition;
+        @Nullable
+        private ExecutorService getter;
+        @Nullable
+        private ExecutorService opener;
+        @Nullable
+        private ExecutorService timeouter;
 
-        Workers() {
-            lock = new StampedLock().asWriteLock();
+        AsyncWorkManager() {
+            state = State.NEW;
+            gets = new LinkedBlockingQueue<>();
+            opens = new LinkedBlockingQueue<>();
+            lock = new ReentrantLock();
+            closedOrTaskAddedCondition = lock.newCondition();
         }
 
-        Executor getter() {
-            if (getter == null) {
-                lock.lock();
-                try {
-                    if (getter == null) {
-                        getter = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
-                    }
-                } finally {
-                    lock.unlock();
+        void enqueueGet(final Task task) {
+            enqueue(gets, task);
+        }
+
+        void enqueueOpen(final Task task) {
+            enqueue(opens, task);
+        }
+
+        private void enqueue(final Queue<Task> queue, final Task task) {
+            lock.lock();
+            try {
+                if (initWorkersUnlessClosed()) {
+                    queue.add(task);
+                    closedOrTaskAddedCondition.signalAll();
+                    return;
                 }
+            } finally {
+                lock.unlock();
             }
-            return getter;
+            task.failAsClosed();
```

Comment:
Is this line of code outside of the lock for correctness, performance, or style? If style, I will note that I prefer to avoid early returns and would move this into an else block inside the try.  But it's personal preference so I don't insist on it.

Reply:
This is done for correctness: we should avoid completing callbacks while holding locks. It still may happen because a caller holds a lock while calling `DefaultConnectionPool.get`, but the class does the best it can.